Add jenkins dockerbuild pipelines doc (#1238)
GoogleCloudPlatform/click-to-deploy: k8s/jenkins/docs/CONFIGURE_K8S_PLUGIN_IN_JENKINS.md

# Using Kubernetes to scale Jenkins agents

You can integrate Jenkins with Kubernetes to deploy multiple pods as Jenkins agents on demand. This enables your Jenkins deployment to dynamically scale its workload, and to deploy any container solution inside pipelines and build containers.

This page contains instructions for integrating Jenkins with Kubernetes.

## Architecture

> Using the same architecture deployed, Jenkins primary pod will be able to communicate with K8s API and deploy new pods inside the cluster on demand, those pods are ephemeral and they are only be alive during the execution of the job or pipeline.

Reviewer: ```suggestion
Using the same architecture deployed, the Jenkins primary pod will be able to communicate with K8s API and deploy new pods inside the cluster on demand. These pods are ephemeral and will only be alive during the execution of the job or pipeline.
```

> ![Architecture diagram](resources/jenkins-k8s-app-architecture-with-k8s-plugin.png)

Reviewer: The image isn't rendering
resources/ -> ./resources
Fix in other places

Reviewer: Oh nvm this isn't a problem only the `PNG` issue below

- apiGroups: [""]
  resources: ["pods/log"]
  verbs: ["get","list","watch"]
- apiGroups: [""]
  resources: ["events"]
  verbs: ["watch"]
- apiGroups: [""]
  resources: ["secrets"]
  verbs: ["get"]
EOF
kubectl apply -f jenkins-service-role.yaml -n $NAMESPACE
```

3. Create a role binding for Jenkins service account:
```
cat > jenkins-service-role-binding.yaml << EOF
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: $APP_INSTANCE_NAME-rolebinding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: $APP_INSTANCE_NAME-role
subjects:
- kind: ServiceAccount
  name: $APP_INSTANCE_NAME-serviceaccount
EOF
kubectl apply -f jenkins-service-role-binding.yaml -n $NAMESPACE
```

## Configure Kubernetes plugin

1. Install [Jenkins Kubernetes Plugin](https://plugins.jenkins.io/kubernetes/).

2. Add the service account token to Jenkins, first get the token using the next command:
```
# Service Account Token
kubectl get secret $(kubectl get sa $APP_INSTANCE_NAME-serviceaccount \
-n $NAMESPACE -o jsonpath="{.secrets[0].name}") \
-n $NAMESPACE -o jsonpath="{.data.token}" | base64 --decode
```
3. Create a secret text in Jenkins with the token:

> ![Service account token](resources/configure-jenkins-k8s-plugin1.png)

Reviewer: This file is named with `PNG` not `png`

4. Get the CA certificate and add it to the plugin configuration:
```
# CA Certificate
kubectl get secret $(kubectl get sa $APP_INSTANCE_NAME-serviceaccount \
-n $NAMESPACE -o jsonpath="{.secrets[0].name}") \
-n $NAMESPACE -o jsonpath={.data.'ca\.crt'} | base64 --decode
```
![Add CA certificate](resources/configure-jenkins-k8s-plugin2.png)

5. Get the jenkins service pod ip address and add it to the plugin configuration as Jenkins Url.
```
echo http://$(kubectl get svc $APP_INSTANCE_NAME-jenkins-ui -n $NAMESPACE -o jsonpath="{.spec.clusterIP}"):8080
```

> 6. Since the agent connector service is exposed as ClusterIP is required to use the option Jenkins tunnel in the plugin, you can get the internal service name with the next command:

Reviewer: ```suggestion
6. The agent connector is exposed service as a ClusterIP. You can get the internal service name with the next command:
```
